dirty_cat/gamma_poisson_factorization.py

```diff
@@ -0,0 +1,600 @@
+"""
+Online Gamma-Poisson factorization of string arrays.
+The principle is as follows:
+    1. Given an input string array X, we build its bag-of-n-grams
+       representation V (n_samples, vocab_size).
+    2. Instead of using the n-grams counts as encodings, we look for low-
+       dimensional representations by modeling n-grams counts as linear
+       combinations of topics V = HW, with W (n_topics, vocab_size) the topics
+       and H (n_samples, n_topics) the associated activations.
+    3. Assuming that n-grams counts follow a Poisson law, we fit H and W to
+       maximize the likelihood of the data, with a Gamma prior for the
+       activations H to induce sparsity.
+    4. In practice, this is equivalent to a non-negative matrix factorization
+       with the Kullback-Leibler divergence as loss, and a Gamma prior on H.
+       We thus optimize H and W with the multiplicative update method.
+"""
+import numpy as np
+from distutils.version import LooseVersion
+from scipy import sparse
+import sklearn
+from sklearn.utils import check_random_state, gen_batches
+from sklearn.utils.extmath import row_norms, safe_sparse_dot
+from sklearn.base import BaseEstimator, TransformerMixin
+from sklearn.feature_extraction.text import CountVectorizer, HashingVectorizer
+from sklearn.cluster import KMeans
+from sklearn.neighbors import NearestNeighbors
+
+if LooseVersion(sklearn.__version__) < LooseVersion('0.22'):
+    from sklearn.cluster.k_means_ import _k_init
+elif LooseVersion(sklearn.__version__) < LooseVersion('0.24'):
+    from sklearn.cluster._kmeans import _k_init
+else:
+    from sklearn.cluster._kmeans import kmeans_plusplus
+
+if LooseVersion(sklearn.__version__) < LooseVersion('0.22'):
+    from sklearn.decomposition.nmf import _beta_divergence
+else:
+    from sklearn.decomposition._nmf import _beta_divergence
+
+
+class OnlineGammaPoissonFactorization(BaseEstimator, TransformerMixin):
+    """
+    Online Gamma-Poisson Factorization by minimizing the
+    Kullback-Leibler divergence.
+
+    Parameters
+    ----------
+
+    n_topics : int, default=10
+        Number of topics of the matrix factorization.
+
+    batch_size : int, default=512
+        Number of samples per batch.
+
+    gamma_shape_prior : float, default=1.1
+        Shape parameter for the Gamma prior distribution.
+
+    gamma_scale_prior : float, default=1.0
+        Scale parameter for the Gamma prior distribution.
+
+    rho : float, default=0.95
+        Weight parameter for the update of the W matrix.
+    
+    rescale_rho : boolean, default=False
```

Comment:
If we are keeping to sklearn's naming conventions, then we can use bool instead of boolean.